stdlib/random.gr

```diff
@@ -0,0 +1,176 @@
+import WasiRandom from "sys/random"
+import Result from "result"
+import Int32 from "int32"
+import Int64 from "int64"
+import WasmI32 from "runtime/unsafe/wasmi32"
+import WasmI64 from "runtime/unsafe/wasmi64"
+import Memory from "runtime/unsafe/memory"
+import DS from "runtime/dataStructures"
+
+record Random {
+  seed: Int64,
+  mut counter: Int64,
+  mut initialized: Bool,
+}
+
+let incCounter = random => {
+  random.counter = Int64.incr(random.counter)
+}
+
+// https://arxiv.org/pdf/2004.06278v3.pdf
+@disableGC
+let rec squares = (ctr: Int64, key: Int64) => {
+  // Need to implement with @disableGC because
+  // overflows are part of this algorithm
+  let ctrArg = ctr
+  let keyArg = key
+  let ctr = DS.loadInt64(WasmI32.fromGrain(ctr))
+  let key = DS.loadInt64(WasmI32.fromGrain(key))
+  let mut x = WasmI64.mul(ctr, key)
+  let mut y = x
+  let mut z = WasmI64.add(y, key)
+  // round 1
+  x = WasmI64.add(WasmI64.mul(x, x), y)
+  x = WasmI64.or(WasmI64.shrU(x, 32N), WasmI64.shl(x, 32N))
+  // round 2
+  x = WasmI64.add(WasmI64.mul(x, x), z)
+  x = WasmI64.or(WasmI64.shrU(x, 32N), WasmI64.shl(x, 32N))
+  // round 3
+  x = WasmI64.add(WasmI64.mul(x, x), y)
+  x = WasmI64.or(WasmI64.shrU(x, 32N), WasmI64.shl(x, 32N))
+  let ret = WasmI32.wrapI64(
+    WasmI64.shrU(WasmI64.add(WasmI64.mul(x, x), z), 32N)
+  )
+  let ret = WasmI32.toGrain(DS.newInt32(ret)): Int32
+  Memory.decRef(WasmI32.fromGrain(ctrArg))
```

Comment:
Why decref'ing `ctrArg` and not `ctr`?

Reply:
Because `ctrArg` is the pointer to the (wrapped i64) argument, and `ctr` is the actual i64 it holds. We want to decref the pointer. I'm updating this now to use `@unsafe` anyway, though, so it's moot.